### ISSUE TITLE
0000099: error: unable to read u32 leb128: stack pointer index

### ISSUE THREAD

**Issue by MarkSwanson** [WebAssembly/wabt#838]
I'm compiling Rust code to wasm, and I'm getting the leb128 issue with wasm-validate.

Rust code:

#[no_mangle]
pub extern "C" fn test(n: i32) -> i32 {
    n
}

I'm using the latest wabt master branch as of today, and a Rust May 9, 2018 nightly.
$ cargo build --release --target wasm32-unknown-unknown
$ wasm-validate target/wasm32-unknown-unknown/release/test.wasm

Has LLVM perhaps been updated to use a different format? Maybe wabt is expecting an older format?

**Comment by binji** [comment]
I chatted with @sbc100 about this, and we're wondering if you're using an older version of wabt. The newest versions should be expecting a linking metadata version which is likely provided by the generated wasm file.

**Comment by sbc100** [comment]
I can take a look at this one.

**Comment by sbc100** [comment]
Looks like then new version field is not being added to the linking metadata.  It was recently added so its not surprising.   Also, the issue was being mis-reported due to a bug: https://github.com/WebAssembly/wabt/pull/841.  Oops.

After that change land the error message is more useful:
```
/bin/wasm-objdump ~/Downloads/test.wasm -h

test.wasm:	file format wasm 0x1
*ERROR*: @0x00000097: invalid linking metadata version: 3
```

**Comment by MarkSwanson** [comment]
Thanks for looking into this.
Do you think the correct thing to do next would be for me (or someone) to log a bug with llvm / LLD?

**Comment by binji** [comment]
@sbc100 wrote the wasm implementation of LLD so I imagine he will fix it.

I did think of another thing to change in wabt though; currently `wasm-validate` is failing to parse the binary because it hits the `linking` custom section and it doesn't match. But validation never relies on a custom section, they always should be ignored. You can pass `--ignore-custom-section-errors` to have it not do this, but it should probably be the default (at least for `wasm-validate` -- for `wasm2wat`, I'm not sure).

**Comment by sbc100** [comment]
I suspect that the compiler that generated that object file just needs to be updated to a version that writes the newer metadata format.  Do you know how that object was generated?  By which llvm version exactly?

I don't think `--ignore-custom-section-errors` should be the default.  If you have old/invalid metadata you probably want to catch it early.  Thats the point of wasm-validate right?

**Comment by MarkSwanson** [comment]
I generated the object like this:
cargo build --release wasm32-unknown-unknown
(verbose mode doesn't provide much more relevant info than that...)

My Rust nightly-2018-05-09 uses LLVM 6. Tracking issue:
https://github.com/rust-lang/rust/issues/47683
LLVM 6.0.0 was merged into Rust on Jan 28, 2018.

I note that there is a tracking issue to upgrade Rust to use LLVM 7:
https://github.com/rust-lang/rust/issues/50543

I also note the changelog for LLVM 7 contains a number of wasm-related LLD enhancements.
So... perhaps this will just work when Rust upgrades to LLVM 7?
